🍒[5.7] Remove the need for `convenience` keyword for delegating actor initializers

kavon opened the pull request:
Cherry-pick of https://github.com/apple/swift/pull/41083 and https://github.com/apple/swift/pull/59828

This is part of SE-327 and includes new warnings for the `convenience` keyword appearing and has a fix-it to delete it. Those warnings turn into an error in Swift 6.

Description: As part of SE-327, the `convenience` keyword is no longer required for delegating actor initializers (the presence of it becomes an error in Swift 6). This means the initializers work similarly to structs, e.g., a non-delegating init can now appear in the extension of an actor.
Risk: Medium. This PR introduces an ABI break for any libraries that had a public `actor` init. Specifically, if the library was compiled with Swift 5.6 or 5.5, any program compiled with optimizations enabled and linked against that library will be dependent on the initializer's delegating status remaining fixed. For example, if the library authors change the implementation to no-longer delegate, the Swift <5.7 program will need to be recompiled, or else there will be a missing symbol error. For Swift 5.7+ we now specifically guarantee that changing the implementation of the init to delegate or not won't break existing programs.
Importance: High. The longer we wait to make this ABI break, the worse it will become.
Review by: Doug Gregor
Testing: PR testing
Original PR: https://github.com/apple/swift/pull/41083 https://github.com/apple/swift/pull/59828
Radar: rdar://87567878

kavon committed:
Remove the need for `convenience` on actor inits to delegate.

This is possible because actors do not support inheritance. There
is one specific exception to that rule, which is that an actor
can inherit from `NSObject` just to support ObjC interop.

This means an actor is effectively a final class.

resolves rdar://87568153

kavon committed:
break request-evaluator cycle when synthesizing actor init

When synthesizing the default initializer for an actor, we'd sometimes hit
a cycle when that initializer needs to chain to NSObject.init. The cycle
only happens because we ask if the initializer we're trying to synthesize
is a convenience init in a scenario which only applies to non-final classes.

Since all actors are effectively "final" classes, it's valid to workaround the
cycle by only asking that initializer question for non-final classes, thus
breaking the cycle.

kavon committed:
make sure all actor kinds, including distributed, are "semantically final"

kavon committed:
add test to catch regressions in ABI stability of actor inits

it should not matter whether an initializer is delegating or not when
importing a library. this test helps make sure that stays true.

kavon commented:
@swift-ci please test

kavon committed:
prevent serialization of allocating inits for actors

A public designated initializer of a class would have its allocating
entry-point serialized in the module, meaning with `-O` that entry-point
can get inlined into programs linking against that module. Once that
entry-point is inlined, the program will _require_ that it remain non-delegating,
because it will depend on the 2nd entry-point (for actual initializing) to be in the
library.

As a result of this change, public initializers of an actor should be resilient in a library,
whether their underlying implementation is delegating or not.

kavon commented:
@swift-ci please test

kavon commented:
@swift-ci please test

DougGregor reviewed (changes requested):
I want us to explore the ABI issue in more detail before we commit to this.

kavon committed:
prevent extensions from having designated inits for actors

When I began removing `convenience` from actor inits, I thought it would
be ok for an actor's extensions to be non-delegating. But that's wrong,
because the actor could be in a resilient module and still have its
properties change between stored and computed, meaning that there isn't
a practical way to allow this, unless if the extension and the actor
are in the same file.

For now, just re-ban this behavior before anybody notices :)

kavon committed:
add objc coverage to actor_init_abi test

kavon commented:
@swift-ci please test and merge

kavon commented:
@swift-ci please test and merge

forgot to add a `REQUIRES: objc_interop`

kavon commented:
@swift-ci please test
